acvm/src/compiler/transformers/csat.rs

```diff
@@ -283,24 +331,32 @@ impl CSatTransformer {
         }
 
         // 2. Create Intermediate variables for the multiplication gates
+        let mut remaining_mul_terms = Vec::with_capacity(gate.mul_terms.len());
         for mul_term in gate.mul_terms.clone().into_iter() {
-            let mut intermediate_gate = Expression::default();
-
-            // Push mul term into the gate
-            intermediate_gate.mul_terms.push(mul_term);
-            // Get an intermediate variable which squashes the multiplication term
-            let inter_var = Self::get_or_create_intermediate_vars(
-                intermediate_variables,
-                intermediate_gate,
-                num_witness,
-            );
-
-            // Add intermediate variable as a part of the fan-in for the original gate
-            gate.linear_combinations.push(inter_var);
+            if self.solvable_witness.contains(&mul_term.1)
+                && self.solvable_witness.contains(&mul_term.2)
+            {
+                let mut intermediate_gate = Expression::default();
+
+                // Push mul term into the gate
+                intermediate_gate.mul_terms.push(mul_term);
+                // Get an intermediate variable which squashes the multiplication term
+                let inter_var = Self::get_or_create_intermediate_vars(
+                    intermediate_variables,
+                    intermediate_gate,
+                    num_witness,
+                );
+
+                // Add intermediate variable as a part of the fan-in for the original gate
+                gate.linear_combinations.push(inter_var);
+                self.mark_solvable(inter_var.1);
+            } else {
+                remaining_mul_terms.push(mul_term);
+            }
         }
 
         // Remove all of the mul terms as we have intermediate variables to represent them now
```

Comment:
This comment is now stale and should be updated.

Reply:
Done